internal/queue/mq.go

```diff
@@ -43,69 +43,53 @@ func (mq *MessageQueue) Write(msg interface{}) error {
 	return nil
 }
 
-// Read will read a value from the queue if available, otherwise return an error.
-func (mq *MessageQueue) Read() (interface{}, error) {
+// Dequeue will read a value from the queue and remove it. If the queue
+// is empty, this will block until the queue is closed or a value gets enqueued.
+func (mq *MessageQueue) Dequeue() (interface{}, error) {
```

Comment:
If `Dequeue` is called on an empty queue, wouldn't it block `Enqueue`?

Reply:
If it's empty we'll end up in the condition variables Wait() method which atomically unlocks the lock and goes to sleep. It'll get woken up when it gets signaled (from something getting enqueued or the queue getting closed)

Reply:
ah, the `c: sync.NewCond(m)` part... got it, that makes sense.

Reply:
do we have a test for this scenario by the way?

Reply:
Not that tests this explicitly, let me add

Reply:
@anmaxvl Added!